nvflare/app_common/tie/connector.py

```diff
@@ -0,0 +1,394 @@
+
+        # start the monitor in a separate daemon thread!
+        t = threading.Thread(target=self._monitor, args=(fl_ctx, connector_stopped_cb), daemon=True)
+        t.start()
+
+    def start_applet(self, app_ctx: dict, fl_ctx: FLContext):
+        """Start the applet set to the connector.
+        If self.in_process is True, then the applet will be started in a separate thread.
+        If self.in_process is False, then the applet will be started in a separate process.
+
+        Args:
+            app_ctx: the contextual info for running the applet
+            fl_ctx: FL context
+
+        Returns: None
+
+        """
+        if not self.applet:
+            raise RuntimeError("applet has not been set!")
+
+        engine = fl_ctx.get_engine()
+        workspace = engine.get_workspace()
+        job_id = fl_ctx.get_job_id()
+
+        if self.applet_env == Constant.APPLET_ENV_SELF:
+            self.logger.info("starting applet by itself")
+            self.applet.start(app_ctx)
+            return
+
+        starter = _AppletStarter(self.applet, self.in_process, workspace, job_id)
+        if self.in_process:
+            self.logger.info("starting applet in another thread")
+            t = threading.Thread(
+                target=starter.start,
+                args=(app_ctx,),
+                daemon=True,
+                name="applet",
+            )
+            t.start()
+            if not starter.started:
+                self.logger.error(f"cannot start applet: {starter.error}")
+                raise RuntimeError(starter.error)
+            self.starter = starter
+        else:
+            # start as a separate local process
+            self.logger.info("starting applet in another process")
+            self.process = multiprocessing.Process(
+                target=starter.start,
+                args=(app_ctx,),
+                daemon=True,
+                name="applet",
+            )
+            self.process.start()
+
+    def stop_applet(self, timeout=0.0):
+        """Stop the running of the applet
+
+        Returns: None
+
+        """
+        if self.applet_env == Constant.APPLET_ENV_SELF:
+            self.applet.stop(timeout)
+            return
+
+        if self.in_process:
+            applet = self.applet
+            self.applet = None
+            if applet:
+                applet.stop(timeout)
+        else:
+            p = self.process
+            self.process = None
+            if p:
+                p.kill()
+
+    def is_applet_stopped(self) -> (bool, int):
+        """Check whether the applet is already stopped
+
+        Returns: a tuple of (whether the applet is stopped, exit code)
+
+        """
+        if self.applet_env == Constant.APPLET_ENV_SELF:
+            if self.applet:
+                return self.applet.is_stopped()
+            else:
+                self.logger.warning("applet is not set with the connector")
+                return True, 0
```

Comment:
should we have a different exit code other than 0?
maybe None or some other NVFlare-defined number?

Reply:
This happens when is_applet_stopped is called before the applet is started (a logic error) or after the applet is stopped, which could happen but harmless.  Don't think we need a special code for this because the warning log message can tell us.

The use of the applet-related methods can not be random - they are called in specific order by other components (controller or executor) to manage its life cycle.